### PR TITLE
Create Block: Generate a block.json file

### PR DESCRIPTION
**Pull request by coreymckrill**
## Description
#22151 suggests auto-generating the block.json file as a way to help block creators be set up for success in getting their blocks added to the block directory.

This adds block.json auto-generation to the `create-block` script during the scaffolding process. The file is pre-populated with the inputs/defaults from the script. The `editorScript`/`editorStyle` properties vary depending on which template is chosen (esnext or es5).

## How has this been tested?
Tested with a local checkout of Gutenberg, running `node path/to/create-block`. Tested with both `esnext` and `es5` values for the `template` parameter.

## Types of changes
* New feature: auto-generated block.json file

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by gziolo** [comment]
Nice one, I'll have a look later this week 👍

File: packages/create-block/lib/init-block-json.js [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+ * Internal dependencies
+ */
+const { info } = require( './log' );
+
+module.exports = async (
+	blockTemplate,
+	{ slug, namespace, title, description, category, dashicon, textdomain }
+) => {
+	const outputFile = join( process.cwd(), slug, 'block.json' );
+	const template = blockTemplate.defaultValues.slug.split( '-' ).shift();
```
Comment (gziolo) [inline packages/create-block/lib/init-block-json.js:18]: This check isn’t reliable in the long run, it’s very likely that a slug without the template name in it is going to be provided in the future.

Instead, we can put both `editorStyle` and `editorScript` inside the template object. Sn override for `es5` can be added here:

https://github.com/WordPress/gutenberg/blob/66e5e5b20c759845866aa172c6a77bdd287ce7b1/packages/create-block/lib/templates.js#L22

The default value can be added here:

https://github.com/WordPress/gutenberg/blob/66e5e5b20c759845866aa172c6a77bdd287ce7b1/packages/create-block/lib/templates.js#L81

The value should be prefixed with `file:`. See more details in block registration RFC:

https://github.com/WordPress/gutenberg/blob/master/docs/rfc/block-registration.md#assets

File: packages/create-block/lib/init-block-json.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+					title,
+					category,
+					icon: dashicon,
+					description,
+					textdomain,
+					supports: {
+						html: false,
+					},
+					editorScript,
+					editorStyle,
```
Comment (gziolo) [inline packages/create-block/lib/init-block-json.js:40]: I think both templates define also `style`.

**Review by gziolo — COMMENTED** [review]
I proposed that we explicitly set values for scripts and styles in the template config. What do you think? Everything else looks good. Thanks for taking care of it.

File: packages/create-block/lib/init-block-json.js (hunk 1 above)
Comment (coreymckrill) [reply packages/create-block/lib/init-block-json.js]: > This check isn’t reliable in the long run

Good point. Updated here: 7d4a923f3

File: packages/create-block/lib/init-block-json.js (hunk 2 above)
Comment (coreymckrill) [reply packages/create-block/lib/init-block-json.js:44]: Added in 7d4a923

File: packages/create-block/lib/templates.js [hunk 3]
```diff
@@ -20,6 +20,9 @@ const predefinedBlockTemplates = {
 			description:
 				'Example block written with ES5 standard and no JSX – no build step required.',
 			wpScripts: false,
+			editorScript: 'file:index.js',
+			editorStyle: 'file:editor.css',
+			style: 'file:style.css',
```
Comment (gziolo) [inline packages/create-block/lib/templates.js:25]:
```suggestion
			editorScript: 'file:./index.js',
			editorStyle: 'file:./editor.css',
			style: 'file:./style.css',
```

File: packages/create-block/lib/templates.js [hunk 4]
```diff
@@ -79,6 +82,9 @@ const getDefaultValues = ( blockTemplate ) => {
 		licenseURI: 'https://www.gnu.org/licenses/gpl-2.0.html',
 		version: '0.1.0',
 		wpScripts: true,
+		editorScript: 'file:build/index.js',
+		editorStyle: 'file:build/index.css',
+		style: 'file:build/style-index.css',
```
Comment (gziolo) [inline packages/create-block/lib/templates.js:87]:
```suggestion
		editorScript: 'file:./build/index.js',
		editorStyle: 'file:./build/index.css',
		style: 'file:./build/style-index.css',
```

**Review by gziolo — APPROVED** [review]
The idea was to use the syntax similar to what npm uses for local packages:

`file:./path/relative/to/block-json/file`

I guess we could skip `./` if that helps. Let me know what you think. I will test it with `register_block_type_from_metadata` helper and merge it tomorrow.

In fact, we could do feature detection in ESNext template and use `register_block_type_from_metadata` when it's defined. I can open PR proposing that tomorrow.

**Comment by coreymckrill** [comment]
> I guess we could skip ./ if that helps.

I don't think it matters much either way, as long as the required syntax is well documented.

**Comment by mkaz** [comment]
Should we update the `index.js` to import the fields from the block.json file?

Right now things like name, title, description, etc.. are duplicated in the two locations.

**Comment by gziolo** [comment]
@mkaz, there are two issues that need to be resolved first:
- JSON imports work only in Node.js or with webpack. In Gutenberg we use a custom plugin that inlines JSON in JavaScript as a plain object. We would need to add the same plugin to the shared Babel preset.
- we miss the logic that wraps translatable fields with i18n functions, it still needs to be added.

In the past, I tried to build a Babel macro that would address both points but it wasn’t as straightforward as you would expect. We were afraid it could be a point of bugs when not configured properly. Related PR: https://github.com/WordPress/gutenberg/pull/16088.

**Comment by gziolo** [comment]
I opened an issue that tracks i18n support for `block.json` in https://github.com/WordPress/gutenberg/issues/23636.

**Comment by mkaz** [comment]
@gziolo Thanks for the update, I've added info to the Create a Block tutorial with a note for the block.json part.

Can you review here: https://github.com/WordPress/gutenberg/pull/23654/files#diff-c52b6bb71644b3ce80dd15e4a3e42ac4

**Pull request merged**